workitem/event/event_repository.go

```diff
@@ -0,0 +1,143 @@
+package event
+
+import (
+	"context"
+	"strings"
+
+	"github.com/jinzhu/gorm"
+	errs "github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/fabric8-services/fabric8-wit/account"
+	"github.com/fabric8-services/fabric8-wit/workitem"
+)
+
+// APIStringTypeEvents represent the type of event
+const APIStringTypeEvents = "events"
+
+// Repository encapsulates retrieval of work item events
+type Repository interface {
+	//repository.Exister
+	List(ctx context.Context, wiID uuid.UUID) ([]Event, error)
+}
+
+// NewEventRepository creates a work item event repository based on gorm
+func NewEventRepository(db *gorm.DB) *GormEventRepository {
+	return &GormEventRepository{
+		db:               db,
+		workItemRepo:     workitem.NewWorkItemRepository(db),
+		wiRevisionRepo:   workitem.NewRevisionRepository(db),
+		workItemTypeRepo: workitem.NewWorkItemTypeRepository(db),
+		identityRepo:     account.NewIdentityRepository(db),
+	}
+}
+
+// GormEventRepository represents the Gorm model
+type GormEventRepository struct {
+	db               *gorm.DB
+	workItemRepo     *workitem.GormWorkItemRepository
+	wiRevisionRepo   *workitem.GormRevisionRepository
+	workItemTypeRepo *workitem.GormWorkItemTypeRepository
+	identityRepo     *account.GormIdentityRepository
+}
+
+// List return the events
+func (r *GormEventRepository) List(ctx context.Context, wiID uuid.UUID) ([]Event, error) {
+	revisionList, err := r.wiRevisionRepo.List(ctx, wiID)
+	if err != nil {
+		return nil, errs.Wrapf(err, "error during fetching event list")
+	}
+	if revisionList == nil {
+		return []Event{}, nil
+	}
+	wi, err := r.workItemRepo.LoadByID(ctx, wiID)
+	if err != nil {
+		return nil, errs.Wrapf(err, "error during fetching event list")
+	}
+	wiType, err := r.workItemTypeRepo.LoadByID(ctx, wi.Type)
+	if err != nil {
+		return nil, errs.Wrapf(err, "error during fetching event list")
+	}
+
+	eventList := []Event{}
+	for k := 1; k < len(revisionList); k++ {
+		modifierID, err := r.identityRepo.Load(ctx, revisionList[k].ModifierIdentity)
+		if err != nil {
+			return nil, errs.Wrapf(err, "error during fetching event list")
+		}
+		for fieldName := range wiType.Fields {
+			switch fieldName {
+			case workitem.SystemAssignees, workitem.SystemLabels:
```

Comment:
@baijum @DhritiShikhar what is so special about these two fields?

Reply:
The value is in array form

Reply:
@baijum okay. To be more generic I suggest that you don't switch on the `fieldName` but on the field's `Type`. Something like this:

```go
for fieldName, field := range wiType.Fields {
    // ...
    switch fieldType := field.Type.(type) {
    case workitem.ListType:
        switch fieldType.ComponentType.Kind {
        case workitem.KindString:
            // this is what assignees and labels are
        }
    case workitem.EnumType: // This would be for state changes
    case workitem.SimpleType: // This is for all the rest
    default: // This is an error
    }
}
```

@aslakknutsen do you agree?

Reply:
@kwk Yea. As little code as possible should care about the 'Field by name', it should care about the type. When we lock our selves down to File names everywhere we're in a world of pain when 'user uploading custom WIT's starts happening in n months.